hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/EightToSevenDowngradeHandler.java

```diff
@@ -92,4 +115,52 @@ private static void downgradePartitionFields(HoodieWriteConfig config, HoodieEng
       tablePropsToAdd.put(HoodieTableConfig.PARTITION_FIELDS, tableConfig.getPartitionFieldProp());
     }
   }
+
+  static void downgradeMetadataPartitions(HoodieEngineContext context,
+                                          HoodieStorage hoodieStorage,
+                                          HoodieTableMetaClient mdtMetaClient,
+                                          Map<ConfigProperty, String> tablePropsToAdd) {
+    // Fetch metadata partition paths.
+    List<String> metadataPartitions = FSUtils.getAllPartitionPaths(
+        context,
+        hoodieStorage,
+        mdtMetaClient.getBasePath(),
+        true);
+
+    // Delete partitions.
+    List<String> validPartitionPaths = deleteMetadataPartition(context, mdtMetaClient, metadataPartitions);
+
+    // Clean the configuration.
+    tablePropsToAdd.put(TABLE_METADATA_PARTITIONS, String.join(",", validPartitionPaths));
+  }
+
+  static List<String> deleteMetadataPartition(HoodieEngineContext context,
+                                              HoodieTableMetaClient mdtMetaClient,
+                                              List<String> metadataPartitions) {
+    List<String> validPartitionPaths = new ArrayList<>();
+
+    for (String partitionPath : metadataPartitions) {
```

Comment:
possible to use `metadataPartitions.stream().map()` to do the deletion and get `validPartitionPaths` in one shot?

Reply:
I think there definitely be a way to do that; but essentially it will use if-else statement inside. I did not find simple one that can do both without if-else statement

Reply:
you can return `Option.empty` or `Option.of(obj)` in the map function and use `.filter` to filter out empty ones corresponding to the deleted partitions.

Reply:
TBH, the readability of using Option is low. I can understand the for-loop by 1 second, but it will take me much more time to figure out Stream api.

Reply:
Using `.map` is actually much cleaner and easier to understand:
```
return metadataPartitions.stream().map(partitionPath -> {
     if (FUNCTIONAL_INDEX == fromPartitionPath(partitionPath)
          || SECONDARY_INDEX == fromPartitionPath(partitionPath)
          || PARTITION_STATS == fromPartitionPath(partitionPath)) {
        HoodieTableMetadataUtil.deleteMetadataTablePartition(
            mdtMetaClient, context, partitionPath, true);
        return Option.empty();
      }
      return partitionPath;
})
.filter(Option::isPresent)
.map(e -> e.get())
.collect();
```